core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -926,30 +933,48 @@ private[spark] class BlockManager(
         } else if (level.useDisk && diskStore.contains(blockId)) {
-          val diskData = diskStore.getBytes(blockId)
-          val iterToReturn: Iterator[Any] = {
-            if (level.deserialized) {
-              val diskValues = serializerManager.dataDeserializeStream(
-                blockId,
-                diskData.toInputStream())(info.classTag)
-              maybeCacheDiskValuesInMemory(info, blockId, level, diskValues)
-            } else {
-              val stream = maybeCacheDiskBytesInMemory(info, blockId, level, diskData)
-                .map { _.toInputStream(dispose = false) }
-                .getOrElse { diskData.toInputStream() }
-              serializerManager.dataDeserializeStream(blockId, stream)(info.classTag)
+          try {
+            val diskData = diskStore.getBytes(blockId)
+            val iterToReturn: Iterator[Any] = {
+              if (level.deserialized) {
+                val diskValues = serializerManager.dataDeserializeStream(
+                  blockId,
+                  diskData.toInputStream())(info.classTag)
+                maybeCacheDiskValuesInMemory(info, blockId, level, diskValues)
+              } else {
+                val stream = maybeCacheDiskBytesInMemory(info, blockId, level, diskData)
+                  .map { _.toInputStream(dispose = false) }
+                  .getOrElse { diskData.toInputStream() }
+                serializerManager.dataDeserializeStream(blockId, stream)(info.classTag)
+              }
             }
+            val ci = CompletionIterator[Any, Iterator[Any]](iterToReturn, {
+              releaseLockAndDispose(blockId, diskData, taskContext)
+            })
+            Some(new BlockResult(ci, DataReadMethod.Disk, info.size))
+          } catch {
+            case ex: KryoException if ex.getCause.isInstanceOf[IOException] =>
+              // We need to have clear error message to catch environmental problems easily.
+              // Further details: https://issues.apache.org/jira/browse/SPARK-37710
+              processKryoException(ex, blockId)
+              throw ex
           }
-          val ci = CompletionIterator[Any, Iterator[Any]](iterToReturn, {
-            releaseLockAndDispose(blockId, diskData, taskContext)
-          })
-          Some(new BlockResult(ci, DataReadMethod.Disk, info.size))
         } else {
           handleLocalReadFailure(blockId)
         }
     }
   }
 
+  private def processKryoException(ex: KryoException, blockId: BlockId): Unit = {
+    var errorMessage = s"${ex.getMessage}. Please check if environment status is healthy for " +
+        s"disk corruption, network failure (etc). ${blockManagerId.toString} - blockName: $blockId"
+    if (diskBlockManager.containsBlock(blockId)) {
+      val file = diskBlockManager.getFile(blockId)
+      errorMessage = errorMessage + s" - blockDiskPath: ${file.getAbsolutePath}"
+    }
+    logError(errorMessage, ex)
```

Comment:
log and rethrow pattern is anti-pattern - I am fine with a `logDebug` or `logInfo` giving additional info (blockManagerId, blockId, file name) in case this context is lost - the stack trace is already getting logged elsewhere.

Reply:
Patch' s main goal is to provide hint by logging `BlockManager`, `blockId` and `blockPath` for potential disk problems occurring on `Kryo deserialization` flow. I think we just need to log problematic `BlockManager` and `block` info so why using `error` log-level. (`ex` can be removed from `logError` due to logging at `Executor$TaskRunner`.) 
On the other hand, re-throwing exception due to avoid breaking current execution flow (e.g: `BlockManager.getLocalValues` is called by `TorrentBroadcast.readBroadcastBlock` and it also catches `IOException/NanFatal`, `log` and `re-throw`). Current execution flow is as follows:
```
FileInputStream => Compression API calls => Kryo Deserialization Flow => BlockManager => RDD => MapPartitionsRDD => ResultTask => Task => Executor$TaskRunner...
```

Reply:
To clarify - log exception + rethrow is an antipattern.
I am fine with either augmenting the thrown exception with additional details (typically the better way to handle this), or logging additional details (like blockId, etc) and throwing original exception to be appropriately handled - avoiding duplicate stacktraces across log file.

The latest change addresses this part, so resolving the conversation.